# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 2:  Lists and For Loops for TVM Problems with many Cashflows

Use __variables__, __lists__ and __for loops__ where appropriate! 

You have the following savings plan:

- 1,000 USD today
- 1,200 USD in 1 year 
- 1,500 USD in 2 years
- 2,000 USD in 3 years
- 2,500 USD in 4 years 
- 3,000 USD in 5 years
- 3,500 USD in 6 years

1. __Create a list__ and __store__ the list in memory!

In [1]:
cf = [1000, 1200, 1500,2000, 2500,3000,3500]

In [2]:
cf

[1000, 1200, 1500, 2000, 2500, 3000, 3500]

(result: [1000, 1200, 1500, 2000, 2500, 3000, 3500])

2. __Index__ the list for the following elements!  

Today´s saving

In [3]:
cf[0]

1000

Saving in year 3

In [4]:
cf[2]

1500

The second element of the list

In [5]:
cf[1]

1200

Final year´s saving

In [6]:
cf[-1]


3500

3. __Iterate__ over the list and __print__ each and every element.

In [8]:
for i in cf:
    print(i)

1000
1200
1500
2000
2500
3000
3500


4. Calculate the __sum__ of your savings. Use a __for loop__ and print intermediate results!

In [9]:
total=0
for i in cf:
    total+=i
    print(total)

1000
2200
3700
5700
8200
11200
14700


(result: 14,700 USD)

5. Now, use two __Python built-in function__ to calculate the sum and the number of elements. 

In [10]:
#sum
sum(cf)

14700

(result: 14,700 USD)

In [11]:
#lenght
len(cf)

7

(result: 7)

6. __Print__ all integers from __0 until 100__. Use __range()__!

In [13]:
print(list(range(100)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]


7. Create and save the list n with all timestamps (0, 1, 2, 3, 4, 5, 6). Use __range()__!

In [16]:
n=list(range(7))

In [17]:
n

[0, 1, 2, 3, 4, 5, 6]

8. __Reverse__ the order of n and __reassign__ n!

In [20]:
n=n[::-1]

(result: [6, 5, 4, 3, 2, 1, 0]

9. Calculate the __future value__ of your account after year 6. Assume an interest rate of 4.0% p.a. 

In [21]:
f=1.04

In [24]:
FV = 0
for i in range(7):
    FV += cf[i] * f**n[i]
    print(FV)

1265.3190184960004
2725.3025013760007
4480.090341376001
6729.818341376001
9433.818341376002
12553.818341376002
16053.818341376002


(result: 16,053.82 USD)

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 200 USD in t1, 
- 300 USD in t2,
- 500 USD in t3, 
- 750 USD in t4,
- 1,000 USD in t5. (each cf at period´s end)<br>


10. The machine costs __2,000 USD__ (Investment in to). Calculate the Project´s __NPV__ and evaluate whether ABC should pursue the project. <br> ABC´s required rate of return (Cost of Capital) is __8%__ p.a. 

In [25]:
cf = [-2000, 200, 300, 500, 750, 1000]
f=1.08

In [26]:
NPV = 0
for i in range(len(cf)):
    NPV += cf[i] / f**(i)
print(NPV)

71.15853841689773


(result: 71.16 USD)

11. Would ABC accept the project with a __required rate of return of 10.0% p.a.__?

In [27]:
f=1.1

In [28]:
NPV = 0
for i in range(len(cf)):
    NPV += cf[i] / f**(i)
print(NPV)

-61.40911885055107


(result: -61.41)

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Exercise 2:  Lists and For Loops for TVM Problems with many Cashflows

Use __variables__, __lists__ and __for loops__ where appropriate! 

You have the following savings plan:

- 1,000 USD today
- 1,200 USD in 1 year 
- 1,500 USD in 2 years
- 2,000 USD in 3 years
- 2,500 USD in 4 years 
- 3,000 USD in 5 years
- 3,500 USD in 6 years

1. __Create a list__ and __store__ the list in memory!

In [ ]:
cf = [1000, 1200, 1500, 2000, 2500, 3000, 3500]

In [ ]:
cf

(result: [1000, 1200, 1500, 2000, 2500, 3000, 3500])

2. __Index__ the list for the following elements!  

Today´s saving

In [ ]:
cf[0]

Saving in year 3

In [ ]:
cf[3]

The second element of the list

In [ ]:
cf[1]

Final year´s saving

In [ ]:
cf[-1]

3. __Iterate__ over the list and __print__ each and every element.

In [ ]:
for element in cf:
    print(element)

4. Calculate the __sum__ of your savings. Use a __for loop__ and print intermediate results!

In [ ]:
total = 0
for element in cf:
    total += element
    print(total)
total

(result: 14,700 USD)

5. Now, use two __Python built-in function__ to calculate the sum and the number of elements. 

In [ ]:
#sum
sum(cf)

(result: 14,700 USD)

In [ ]:
#lenght
len(cf)

(result: 7)

6. __Print__ all integers from __0 until 100__. Use __range()__!

In [ ]:
for i in range(101):
    print(i)

7. Create and save the list n with all timestamps (0, 1, 2, 3, 4, 5, 6). Use __range()__!

In [ ]:
n = list(range(7))

In [ ]:
n

8. __Reverse__ the order of n and __reassign__ n!

In [ ]:
n = n[::-1]

In [ ]:
n

(result: [6, 5, 4, 3, 2, 1, 0]

9. Calculate the __future value__ of your account after year 6. Assume an interest rate of 4.0% p.a. 

In [ ]:
r = 0.04
f = 1 + r

In [ ]:
FV = 0
for i in range(len(cf)):
    FV += cf[i] *  f**n[i]
print(FV)

(result: 16,053.82 USD)

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 200 USD in t1, 
- 300 USD in t2,
- 500 USD in t3, 
- 750 USD in t4,
- 1,000 USD in t5. (each cf at period´s end)<br>


10. The machine costs __2,000 USD__ (Investment in to). Calculate the Project´s __NPV__ and evaluate whether ABC should pursue the project. <br> ABC´s required rate of return (Cost of Capital) is __8%__ p.a. 

In [ ]:
cf = [-2000, 200, 300, 500, 750, 1000]
r = 0.08
f = 1 + r

In [ ]:
NPV = 0
for i in range(len(cf)):
    NPV += cf[i] / f**(i)
print(NPV)

(result: 71.16 USD -> accept)

11. Would ABC accept the project with a __required rate of return of 10.0% p.a.__?

In [ ]:
cf = [-2000, 200, 300, 500, 750, 1000]
r = 0.10
f = 1 + r

In [ ]:
NPV = 0
for i in range(len(cf)):
    NPV += cf[i] / f**(i)
print(NPV)

(result: -61.41 -> reject)